src/sidebar/services/groups.js

```diff
@@ -163,164 +163,144 @@ function groups(
+    // the group may not be in the results returned by group.list,
+    // profile.groups, or the direct-linked group.
+    const directLinkedAnnId = store.getState().directLinkedAnnotationId;
+    let directLinkedAnnApi = null;
+    if (directLinkedAnnId) {
+      directLinkedAnnApi = api.annotation
+        .get({ id: directLinkedAnnId })
+        .catch(() => {
+          // If the annotation does not exist or the user doesn't have permission.
+          return null;
+        });
+    }
 
-    // Step 1: Get the URI of the active document, so we can fetch groups
-    // associated with that document.
-    return uri
-      .then(uri => {
-        // Step 2: Concurrently fetch the groups the user is a member of,
-        // the groups associated with the current document and the annotation
-        // or group that was direct-linked (if any).
-        if (authority) {
-          params.authority = authority;
-        }
-        if (uri) {
-          params.document_uri = uri;
+    // If there is a direct-linked group, add an API request to get that
+    // particular group since it may not be in the results returned by
+    // group.list or profile.groups.
+    const directLinkedGroupId = store.getState().directLinkedGroupId;
+    let directLinkedGroupApi = null;
+    if (directLinkedGroupId) {
+      directLinkedGroupApi = fetchGroup({
+        id: directLinkedGroupId,
+        expand: params.expand,
+      }).then(group => {
+        // If the group does not exist or the user doesn't have permission.
+        if (group === null) {
+          store.setDirectLinkedGroupFetchFailed();
+        } else {
+          store.clearDirectLinkedGroupFetchFailed();
         }
-        documentUri = uri;
+        return group;
+      });
+    }
 
-        const profileGroupsApi = api.profile.groups.read({
+    const [
+      myGroups,
+      featuredGroups,
+      token,
+      directLinkedAnn,
+      directLinkedGroup,
+    ] = await Promise.all([
+      api.profile.groups.read({ expand: params.expand }),
+      api.groups.list(params),
+      auth.tokenGetter(),
+      directLinkedAnnApi,
+      directLinkedGroupApi,
+    ]);
+
+    // Step 3. Add the direct-linked group to the list of featured groups,
+    // and if there was a direct-linked annotation, fetch its group if we
+    // don't already have it.
+
+    // If there is a direct-linked group, add it to the featured groups list.
+    if (
+      directLinkedGroup &&
+      !featuredGroups.some(g => g.id === directLinkedGroup.id)
+    ) {
+      featuredGroups.push(directLinkedGroup);
+    }
+
+    // If there's a direct-linked annotation it may require an extra API call
+    // to fetch its group.
+    let directLinkedAnnotationGroupId = null;
+    if (directLinkedAnn) {
+      // Set the directLinkedAnnotationGroupId to be used later in
+      // the filterGroups method.
+      directLinkedAnnotationGroupId = directLinkedAnn.group;
+
+      const directLinkedAnnGroup = myGroups
```

Comment:
While reviewing I had to really think about why this was checking for the group in both lists. Maybe this was just me having a durp moment but I was thinking that might indicate that a comment might be in order here. I was really struggling with the why part of this which is actually in the comment after this line.

`// Check to see if the group has already been fetched.`

Reply:
I'll move the comment below up to make this clearer.